src/Simplex/Chat.hs

```diff
@@ -1655,6 +1655,43 @@ processChatCommand' vr = \case
     case conn'_ of
       Just conn' -> pure $ CRConnectionIncognitoUpdated user conn'
       Nothing -> throwChatError CEConnectionIncognitoChangeProhibited
+  APISetConnectionUserId connId newUserId -> withUser $ \user@User {userId} -> do
+    cfg <- asks config
+    newUser <- privateGetUser newUserId
+    conn' <- withFastStore $ \db -> do
+      conn <- getPendingContactConnection db userId connId
+      newUserServers <- liftIO $ getServers db cfg newUser SPSMP
+      pure $ checkConnection conn newUserServers
+    case conn' of
+      (True, Just conn) -> do
+        -- Safe to update connection
+        withAgent $ \a -> changeConnectionUser a (aUserId user) (aConnId' conn) (aUserId newUser)
+        _ <- withFastStore' $ \db -> updatePCCUser db userId conn newUserId
+        pure $ CRConnectionUserChanged user conn newUser
+      (False, Just conn) -> do
+        -- Servers do not match, create new connection
+        subMode <- chatReadVar subscriptionMode
+        (newConnId, cReq) <- withAgent $ \a -> createConnection a (aUserId user) True SCMInvitation Nothing IKPQOn subMode
+        newConn <- withFastStore' $ \db -> do
+          deleteConnectionRecord db user connId
+          createDirectConnection db newUser newConnId cReq ConnNew Nothing subMode initialChatVersion PQSupportOn
+        withAgent $ \a -> deleteConnectionAsync a False (aConnId' conn)
+        pure $ CRInvitation user cReq newConn
+      _ -> throwChatError CEInvalidConnReq
+    where
+      getServers :: (ProtocolTypeI p, UserProtocol p) => DB.Connection -> ChatConfig -> User -> SProtocolType p -> IO (NonEmpty (ProtocolServer p))
+      getServers db cfg user p = L.map (\ServerCfg {server} -> protoServer server) . useServers cfg p <$> getProtocolServers db user
+      checkConnection :: PendingContactConnection -> NonEmpty (ProtocolServer 'PSMP) -> (Bool, Maybe PendingContactConnection)
+      checkConnection conn@PendingContactConnection {pccConnStatus, connReqInv} newUserServers =
+        case (pccConnStatus, connReqInv) of
+          (ConnNew, Just req) ->
+            let r = connectionData req
+                ConnReqUriData {crSmpQueues = q :| _} = r
+                SMPQueueUri {queueAddress = SMPQueueAddress {smpServer}} = q
+             in (smpServer `elem` newUserServers, Just conn)
+          _ -> (False, Nothing)
```

Comment:
this currently creates some indirection and may turn out better if we check pccConnStatus, connReqInv outside and simply throw error based on that, then in good path check servers and do stuff depending on whether server has to change (so this function would only check server is in list)

Reply:
also would be more straightforward to separate reading servers into the good branch then - as we don't need them if link is in bad status